jackson2 module

stevenschlansker opened the pull request:
adds `@Json` type qualifier

qualidafial commented on jackson2/src/main/java/org/jdbi/v3/jackson2/Jackson2Config.java:
```diff
@@ -0,0 +1,51 @@
+ */
+@Beta
+public class Jackson2Config implements JdbiConfig<Jackson2Config> {
+    private ObjectMapper mapper;
```
Do we want to default this to a `new ObjectMapper()` for basic out of the box functionality?

stevenschlansker replied:
I waffled a bit on that one.  Personally, I always configure my Mappers, so this prevents me from accidentally forgetting to configure it.  It also means whatever mapper caching Jackson does is shared with the rest of the application.  I'm happy to change this if anyone feels strongly, though.

qualidafial replied:
I usually configure the module 2, but for those who don't need to I'd rather have it "just work" out fo the box.

An alternative to having a default is to no-arg constructor to the plugin that uses a `new ObjectMapper()`

TheRealMarnes replied:
I think it feels more in line with how the rest of jdbi works if we default this to a `new`

qualidafial commented on jackson2/src/test/java/org/jdbi/v3/jackson2/TestJackson2Plugin.java:
```diff
@@ -0,0 +1,88 @@
+        @SqlQuery("select whozit from whozits")
+        @Json
+        List<Whozit> select();
+    }
```
I'd also like to see some tests with a `@Json` property of e.g. a bean class.

stevenschlansker commented on jackson2/src/main/java/org/jdbi/v3/jackson2/JsonArgumentFactory.java:
```diff
@@ -0,0 +1,39 @@
+    public Optional<Argument> build(Type type, Object value, ConfigRegistry config) {
+        return Optional.of((p, s, c) -> {
+            try {
+                s.setObject(p, config.get(Jackson2Config.class).getMapper().writeValueAsString(value), Types.OTHER);
```
Agreed.  I only use Postgres and I didn't want to make the initial pass overly complicated.

qualidafial commented on jackson2/src/test/java/org/jdbi/v3/jackson2/TestJackson2Plugin.java:
```diff
@@ -0,0 +1,88 @@
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestJackson2Plugin {
+// tag::jackson2example[]
```
Dropping the whole unit test into the docs doesn't exactly convey how the plugin should be used.

Not a blocker for merge, but I think users would find more value in a conversational "first you add the dep: <example>, then you install the plugin: <example>, then if you annotate your bound properties or SQL objects like this: <example> then code like this Just Works :tm: <example>

stevenschlansker commented:
I addressed all the review comments.

qualidafial commented on jackson2/src/main/java/org/jdbi/v3/jackson2/Jackson2Plugin.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Override
+    public void customizeJdbi(Jdbi jdbi) {
+        if (mapper != null) {
```
We're already checking for non-null in the constructor

TheRealMarnes commented on jackson2/src/main/java/org/jdbi/v3/jackson2/Jackson2Config.java:
```diff
@@ -0,0 +1,51 @@
+
+    @Override
+    public Jackson2Config createCopy() {
+        return new Jackson2Config().setMapper(mapper);
```
perhaps a copy constructor to be consistent with all other Config classes? Would be a shame to divert from the pattern and have to refactor it later anyway.

stevenschlansker replied:
It's all internal and can be changed on a whim with no problem.  Why is the copy constructor important?  I'm experimenting a bit and don't you agree that having named chained setters is usually preferable to long constructors?  (You were advocating *against* constructors with long parameter lists in this very review!)

TheRealMarnes commented on jackson2/src/main/java/org/jdbi/v3/jackson2/Jackson2Plugin.java:
```diff
@@ -0,0 +1,54 @@
+        this(new ObjectMapper());
+    }
+
+    public Jackson2Plugin(ObjectMapper mapper) {
```
kinda feels like this is different from how the other plugins work, delegating all their configuration to Config classes alone instead of having special constructors that will eventually become a burden when more configuration is added

qualidafial replied:
I'm ambivalent on this one.

I do agree that it's a departure from how other plugins work, so maybe it's worth making it consistent for consistency's sake.

qualidafial replied:
@TheRealMarnes by "ugly mess" do you mean overloading constructors for more config parameters? I don't see that happening. The `ObjectMapper` is the focal point of this module so it's kind of special in that sense.

TheRealMarnes replied:
That's what they always say at first :)

But yeah, the sum of _maybe_ wanting to add more in the future because of expansion and the deviation from an established pattern, makes me say we shouldn't do this.

qualidafial replied:
On reflection I think I agree with Marnes on this.

We can always change our mind in the future and put this constructor back, but we can't take it back once it's published API.

stevenschlansker replied:
I never intended to continue adding new constructors if there's any more configuration to do.  I only had this constructor because it really is an integral part of the integration, and currently there's no need for further configuration - so you don't need to know about the config class at all.  I guess consistency is an argument, but I do think ObjectMapper is integral to this module in a way that e.g. changing null binding or configuring case mapping isn't.

TheRealMarnes replied:
I really don't see this constructor as added value over the config class, especially because we can provide a default anyway...

If we go ahead with splitting jackson out of the feaure itself as discussed further below though, then this constructor would become a necessity: you could have a `new JacksonPlugin()` or `new GsonPlugin()`, but would require a `new JsonPlugin(mapperImpl)`.

stevenschlansker replied:
I did this.

qualidafial commented on RELEASE_NOTES:
```diff
@@ -11,6 +11,7 @@
     - ParsedSql can be created with ParsedSql.of(String, ParsedParameters) factory
     - ParsedParameters can be created with ParsedSql.positional(int) and
       ParsedSql.named(List<String>) factories.
+    - @Json type qualifier with Jackson 2.0 bindings
```
Should move this to the beta API section

qualidafial commented on jackson2/pom.xml:
```diff
@@ -0,0 +1,81 @@
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-databind</artifactId>
+            <version>${jackson.version}</version>
```
Should we manage this in the parent POM instead?

stevenschlansker replied:
I'd put it here since no other modules should ever depend on jackson, so there is no reason for them to inherit its management.  Do you think we should put it in the parent anyway?

TheRealMarnes replied:
I think it's good practice to always put dependency versions in the parent regardless of usage location, just to avoid ever getting different versions of the same thing or spreading version management out over multiple places.

stevenschlansker replied:
Right, and that's exactly what I'd imagined here -- someday there will be a Jackson 3 and that will be a new `jackson3` module, and you don't want the `jackson2` dependency management creeping in.  Or someone (foolishly) decides to contribute a `jackson1` module.  So would you consider this such an exception?

TheRealMarnes replied:
Well you could leave it in the parent because there's no (expectation of a) competing version module _yet_, or you could move it inside the module preemptively... The difference though is that with the former you'll notice the necessary change at dev time, while with the latter another jackson version might be added to another module for whatever reason and then there'll be conflicts _at runtime_ for users. It feels like opening the door to an (unlikely) unwelcome surprise. I would prefer it in the parent until there's a competing module, but it's not like I'm gonna make a big deal out of it if you feel strongly and make the decision to not do that ¯\\\_(ツ)\_/¯

qualidafial commented on jackson2/src/main/java/org/jdbi/v3/jackson2/Jackson2Plugin.java:
```diff
@@ -0,0 +1,52 @@
+
+    private final ObjectMapper mapper;
+
+    public Jackson2Plugin() {
```
Should we produce a `META-INF/services` SPI file for this plugin?

TheRealMarnes commented:
I'll just TLDR here: why don't you make the Json mapping itself an impl-agnostic core feature with an ObjectMapper-ish interface, and let Jackson only be a separate module that provides a Plugin that registers the mapper instance? Then users could just as well use Gson or whatever other json mapper impl, anything that can (de)serialize a String and Type.

stevenschlansker commented:
I considered doing that, as well as factoring out the strategy we use to bind Postgres json.  But I was really just trying to get the feature out the door and in use before making it overly general.  Do we think anyone will actually contribute Gson support?  I'm not opposed or disagreeing, it's just that you're asking for the scope of this to increase a lot...

TheRealMarnes commented:
Well if you made the json mapping a core feature with an SPI, adding any backing impl like jackson or gson would be 5 lines of code max... I'm not intent on adding gson so much as I am worried about unnecessarily tying a generic json feature to a single specific library that can't be substituted by another when there's really no need for such a hard dependency.

TheRealMarnes commented:
The change could actually be as simple as taking the core code out of this module into a jdbi3-json, and moving the jackson implementation code to a jdbi3-jackson module with a dependency on jdbi3-json. Then users would add jdbi3-jackson or another impl, or jdbi3-json and roll their own.

stevenschlansker commented:
So if I added a new module `jdbi-annotations`, moved the `@Json` type qualifier there, but left the only implementation as jackson2 -- is that good enough?

stevenschlansker commented:
Ah, I guess it could be a `json` module as well.  We're getting so many modules...

TheRealMarnes commented:
Making jdbi3-json as a separate module is entirely optional, but I think we should start scaling horizontally instead of tacking more and more into core. I don't really see added modules as a burden, there are only so many integrations we can provide.

TheRealMarnes commented:
But yeah, I would make the entire "this data should be handled and mapped as json" part of this PR an agnostic feature on itself (either in core or a new module), and provide the "... by jackson" part of the sentence as an SPI implementation to delegate the String/Type to only.

stevenschlansker commented:
I split the implementation from the interface, but the changes are significant enough that the code needs re-review.

TheRealMarnes commented on core/src/main/java/org/jdbi/v3/core/internal/JdbiClassUtils.java:
```diff
@@ -0,0 +1,29 @@
+        try {
+            Class.forName(klass);
+            return true;
+        } catch (ReflectiveOperationException e) {
```
should be named `ignored` to satisfy the "log or throw but not both nor neither" rule

TheRealMarnes commented:
I'm happy with the code, but I'm wondering if there's a language barrier here. Do we have the same idea of what a copy constructor is? To me it is a constructor like `private X(X other) { this.foo = other.foo; etc}`, is it that to you too? Did you begrudgingly put regular all-arg constructors on the configs because you thought I was asking for that, or did you make the decision not to do copy constructors as I see them and jsut go with a regular constructor?

stevenschlansker commented:
There must be some misunderstanding.  I thought I was doing exactly as you outlined:
https://github.com/jdbi/jdbi/pull/1334/commits/83790281d3b234f42b5df5aace13dcb9930d4638
```diff
+    private JsonConfig(JsonMapperImpl impl) {
+        this.impl = impl;
+    }
```

TheRealMarnes commented:
That explains the friction I felt :)

No actually I'm in favor of copy constructors as I outlined them just now, i.e. not how they are now implemented nor with the setters from before. Working on the other JdbiConfigs showed those are the best solution, and I suspect that you might agree. :)

stevenschlansker commented:
I don't know for sure that I agree, but since I don't feel strongly, I'm willing to take your assurance that it's better :)

stevenschlansker commented:
I'm still somewhat sad about the loss of `Jackson2Module(ObjectMapper)` but it's only a small wart to have to immediately configure it.  I still can't shake the feeling that it's 'essential' configuration in the way most of the other configurables aren't.  But I'm inclined to merge this as is (without it) and maybe add it later if need be.

TheRealMarnes commented on jackson2/src/main/java/org/jdbi/v3/jackson2/Jackson2Config.java:
```diff
@@ -0,0 +1,57 @@
+        this(new ObjectMapper());
+    }
+
+    private Jackson2Config(ObjectMapper mapper) {
```
so yeah let's change this real quick

TheRealMarnes commented on json/src/main/java/org/jdbi/v3/json/JsonConfig.java:
```diff
@@ -0,0 +1,46 @@
+
+    public JsonConfig() {}
+
+    private JsonConfig(JsonMapperImpl impl) {
```
and this one too

stevenschlansker commented:
I think there's a strong argument that exposing `Jackson2Config` is a heavier weight API than another constructor.  But I'm more interested in focusing on shipping this PR right now than worrying about this issue.

TheRealMarnes commented:
I'm curious, what do you see now that wasn't there before? Something about the language barrier, or did eclipse give you the same additional support as intellij when changing the constructors to the new kind?

stevenschlansker commented:
I just totally missed your point.

TheRealMarnes commented:
Hah, ok. Could I have done something better?

TheRealMarnes commented:
And yeah, true, the additional exposed API is much heavier than the constructor, but none of that is redundant. Redundant API exposure feels 1000x heavier to me than new API exposure. But whatever, add the constructor in the future if people ask or something :)

stevenschlansker commented:
I was just frustrated with how many times around this PR is taking and misunderstood.  If you give me a green checkmark all will be forgiven ;)

stevenschlansker commented:
I squished it.

TheRealMarnes commented:
I got the same remark a few days ago from sarxos, but I'm very decidedly with both feet in the "rather safe than sorry" camp. I'll gladly concede something I request in a review if given a firm no, but as long as something is up for debate, I'll keep fighting until I get a hard counterargument. I already feel like crap for having added the MessageFormatTemplateEngine without knowing until now about the integer format issue. In fact I feel stupid not having picked up over all these years that String.format's `%s` is exactly the string formatting tool I've been looking for all this time after all, always having used MessageFormat instead to get around an issue that was purely a misunderstanding.

stevenschlansker commented:
That's the right attitude to have.

TheRealMarnes commented:
Could have just asked me to do that with the button, we wouldn't have to wait for another build 😛

stevenschlansker commented:
Oops, my finger slipped ;)

TheRealMarnes commented:
![](https://66.media.tumblr.com/5eb0d2842f6a2f9cc988b387d2163ac3/tumblr_msz5lmdznG1r1vf9eo1_r1_500.gif)

qualidafial commented on jackson2/src/main/java/org/jdbi/v3/jackson2/JacksonJsonImpl.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @Override
+    public <T> T fromJson(Type type, String json, StatementContext ctx) {
```
With the `Type` parameter, the best this method can promise to return is `Object`

stevenschlansker replied:
You're right, I'll fix this.

stevenschlansker replied:
https://github.com/jdbi/jdbi/pull/1344

qualidafial commented on json/src/main/java/org/jdbi/v3/json/JsonConfig.java:
```diff
@@ -0,0 +1,46 @@
+        this.impl = other.impl;
+    }
+
+    public JsonConfig setImpl(JsonMapperImpl impl) {
```
What do you guys think of renaming the interface to `JsonMapper` and naming this config property `jsonMapper`?

TheRealMarnes replied:
Sounds good to me

qualidafial commented on postgres/src/main/java/org/jdbi/v3/postgres/JsonArgumentFactory.java:
```diff
@@ -0,0 +1,33 @@
+class JsonArgumentFactory implements ArgumentFactory {
+    @Override
+    public Optional<Argument> build(Type type, Object value, ConfigRegistry config) {
+        if (type != Json.class) {
```
What's happening here? `Json.class` isn't the type, is the qualifier. Shouldn't this be an annotation on the factory class declaration?

stevenschlansker replied:
This is the "poor man's SPI" to decouple Postgres support from Json support.
The Json binder matches parameterized types e.g. `@Json Foo`, makes them a Json string, then re-binds them as the type `Json.class` (bit of a hack, but it works!)

It expects the database engine, in this case Postgres, to provide the specific binding strategy.

I guess this could be an ArgumentFactory setting on the config instead if that's clearer.  But we already have an argument factory registry...

qualidafial replied:
It sounds like `QualifiedArgumentFactory` wants to be API.
